src/Workspaces/Core/Portable/Execution/ChecksumTreeCollection.TreeNodes.cs

```diff
@@ -179,6 +247,59 @@ public IChecksumTreeNode GetOrCreateSubTreeNode<TKey>(TKey key)
                 return GetOrCreateChecksumObjectAsync(key, value, kind, valueGetterAsync, cancellationToken);
             }
 
+            protected void AppendChecksumObjects(
+                ImmutableDictionary<Checksum, ChecksumObject>.Builder builder,
+                HashSet<Checksum> checksums,
+                int itemCount,
+                IEnumerable<Checksum> itemChecksums,
+                Func<Checksum, ChecksumObject> itemGetter, CancellationToken cancellationToken)
+            {
+                using (var removed = Creator.CreateList<Checksum>())
+                {
+                    foreach (var checksum in GetChecksums(checksums.Count, checksums, itemCount, itemChecksums))
+                    {
+                        cancellationToken.ThrowIfCancellationRequested();
+
+                        var checksumObject = itemGetter(checksum);
+                        if (checksumObject != null && checksums.Contains(checksum))
+                        {
+                            // found entry
+                            builder[checksum] = checksumObject;
+                            removed.Object.Add(checksum);
+
+                            if (removed.Object.Count == checksums.Count)
+                            {
+                                break;
+                            }
+                        }
+                    }
+
+                    checksums.ExceptWith(removed.Object);
+                }
+            }
+
+            private IEnumerable<Checksum> GetChecksums(int count1, IEnumerable<Checksum> checksums1, int count2, IEnumerable<Checksum> checksums2)
```

Comment:
this return smaller checksum list to loop over. sometime we are asked to find 1000 checksum or something we are asked to find 10 checksum and sub tree something contains only 2-3 items or something it contains 100-200 items.

we will pick which ever collection that has smaller checksum list and loop over that and check 2 sets (one asked, one this subtree owns) whether both contains the checksum.